File: src/udemy_tutorials_cpp/src/add_two_ints_client_no_oop.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"


int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<rclcpp::Node>("add_two_ints_client_no_oop");

    // Create client without OOP
    auto client = node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
    RCLCPP_INFO(node->get_logger(), "AddTwoInts client (no opp) initialized");
    
    // Wait for server to be up
    while (!client->wait_for_service(std::chrono::seconds(1))) {
        RCLCPP_WARN(node->get_logger(), "Service not available, waiting again...");
    }
    RCLCPP_INFO(node->get_logger(), "Service is available");
    // Create request to be send with node
    auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();

    // Set request values
    request->a = 3;
    request->b = 8;

    // Send request to server
    auto future = client->async_send_request(request);
    RCLCPP_INFO(node->get_logger(), "Request sent");
    RCLCPP_INFO(node->get_logger(),
                "  Request: %ld + %ld", request->a, request->b);
    // Wait for response using future
    if (rclcpp::spin_until_future_complete(node, future) == rclcpp::FutureReturnCode::SUCCESS) {
        RCLCPP_INFO(node->get_logger(), "  Sum: %ld", future.get()->sum);
    } else {
        RCLCPP_ERROR(node->get_logger(), "Service call failed");
    } 
    // Shutdown the node
    rclcpp::shutdown();
    return 0;
}

File: src/udemy_tutorials_py/udemy_tutorials_py/add_two_ints_client_no_oop.py
#!/usr/bin/env python3
"""Example to communicate with the add_two_ints service without OOP."""
import sys
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


def main(args=None):
    """Client node without OOP."""
    rclpy.init(args=args)
    # Create a node
    node = Node("add_two_ints_client_no_oop")
    client = node.create_client(AddTwoInts, "add_two_ints")
    node.get_logger().info("Client node has been created.")
    # Wait for the service to be available
    node.get_logger().info("Waiting for the service to be available...")
    while not client.wait_for_service(timeout_sec=1.0):  # Check every second
        node.get_logger().warn("Service not available, waiting again...")
    node.get_logger().info("Service is available.")
    # Create a request
    request = AddTwoInts.Request()
    request.a = 3
    request.b = 4
    # Call the service
    future = client.call_async(request)
    rclpy.spin_until_future_complete(node, future)
    if future.result() is not None:
        node.get_logger().info(f"Result: {future.result().sum}")
    else:
        node.get_logger().error("Service call failed.")
    # Destroy the node explicitly
    node.destroy_node()
    rclpy.shutdown()
    return 0


if __name__ == "__main__":
    sys.exit(main())
